### DISCUSSION TITLE
Require JavaScript files

### DISCUSSION THREAD

**Discussion by PEZ** [BetterThanTomorrow/joyride#114] [Ideas]
It would be cool, and maybe lower the entry bar to be able to require JavaScript files.

A workspace script could be a minimal glue that just require a JavaScript file. And we could even, given that we _can_ require JavaScript files, support running JavaScript files as scripts.

**Comment by borkdude** [comment]
See nbb which already has this

**Reply by PEZ** [reply]
Looking at this now. I find the `load-module` function:

https://github.com/babashka/nbb/blob/79bc68474afa9d79bb037613b36b1ce20b3acb4d/src/nbb/core.cljs#L72

I notice that it is async. Which it probably will need to be in VS Code as well, because of how the APIs work. Will that be a problem, you think?

**Reply by borkdude** [reply]
In joyride it doesn't have to be async since `js/require` is sync

**Reply by PEZ** [reply]
I think that making things async might be necessary to enable Joyride to run on [VS Code Web](https://vscode.dev/). Which would be the ultimate awesomeness.

**Comment by PEZ** [comment]
Looking at how we require things made me tried this without changing anything:

`foo.js`:

```js
exports.hello = () => {
    return 'Hello World!';
}
```

`foo.cljs`:

```clojure
(ns foo
  (:require ["./foo.js" :as js-foo]))

(js-foo/hello)
```

Loading `foo.cljs` in the REPL:

```
"Hello World!"
```

However, we seem to cache the first load of `foo.js`. Re-evaluating the ns form does not give me any new exports from there, or reflect any changes to existing ones.

**Reply by PEZ** [reply]
We might get away with uncaching it: https://nodejs.org/api/modules.html#requirecache

Though, I guess we would want to do this only if the file has changed. Or, rather, whatever scheme is used for cljs files...
